src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolWorkQueue.cs

```diff
@@ -851,6 +861,90 @@ public void TransferLocalWork()
                 ThreadPoolWorkQueue.WorkStealingQueueList.Remove(workStealingQueue);
             }
         }
+
+        [Flags]
+        public enum WorkState
+        {
+            MayHaveLocalWorkItems = 1 << 0,
+            IsProcessingHighPriorityWorkItems = 1 << 1
+        }
+    }
+
+    internal sealed class ThreadPoolTypedWorkItemQueue<T> : IThreadPoolWorkItem
+    {
+        private int _isScheduledForProcessing;
+        private readonly ConcurrentQueue<T> _workItems = new ConcurrentQueue<T>();
+        private readonly Action<T> _callback;
+
+        public ThreadPoolTypedWorkItemQueue(Action<T> callback) => _callback = callback;
+
+        public int Count => _workItems.Count;
+
+        public void Enqueue(T workItem)
+        {
+            BatchEnqueue(workItem);
+            CompleteBatchEnqueue();
+        }
+
+        public void BatchEnqueue(T workItem) => _workItems.Enqueue(workItem);
+        public void CompleteBatchEnqueue() => ScheduleForProcessing();
+
+        private void ScheduleForProcessing()
+        {
+            // Only one thread is requested at a time to avoid over-parallelization. Currently where this type is used, queued
+            // work is expected to be processed at high priority. The implementation could be modified to support different
+            // priorities if necessary.
+            if (Interlocked.CompareExchange(ref _isScheduledForProcessing, 1, 0) == 0)
+            {
+                ThreadPool.UnsafeQueueHighPriorityWorkItemInternal(this);
+            }
+        }
+
+        void IThreadPoolWorkItem.Execute()
+        {
+            Debug.Assert(_isScheduledForProcessing != 0);
+
+            // This change needs to be visible to other threads that may enqueue work items before a work item is attempted to
+            // be dequeued by the current thread. In particular, if an enqueuer queues a work item and does not schedule for
+            // processing, and the current thread is the last thread processing work items, the current thread must see the work
+            // item queued by the enqueuer.
+            _isScheduledForProcessing = 0;
+            Interlocked.MemoryBarrier();
+            if (!_workItems.TryDequeue(out T? workItem))
+            {
+                return;
+            }
+
+            ScheduleForProcessing();
```

Comment:
I haven't understood this line. We queue the object to the thread pool, but we already running it. Why do it again inside the thread pool?

Reply:
To parallelize the work. An explanation of this pattern is here https://github.com/dotnet/runtime/blob/9293aab2379d6a122d0cf5ed350356cb23e4d698/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs#L206.

Reply:
I see. Thanks for the link.